File: AbstractFactory/python-source/rooms/CavesHardRoom.py
from random import random

from rooms.BasementHardRoom import BasementHardRoom
from rooms.HardRoom import HardRoom
from rooms.Room import Room


class CavesHardRoom(HardRoom):
    def generate_loot(self) -> str:
        return 'Loot x4'

    def generate_boss(self) -> str:
        return '400hp'

    def generate_secret(self, previous_room: Room):
        if isinstance(previous_room, CavesHardRoom):
            return random() < 0.5
        if isinstance(previous_room, BasementHardRoom):
            return random() < 0.4
        return random() < 0.3

File: JavaSource/src/main/java/ru/akhcheck/patterns/abstractfactory/rooms/CavesHardRoom.java
package ru.akhcheck.patterns.abstractfactory.rooms;

/** Сложная комната пещер: у неё есть секрет. */
public class CavesHardRoom implements HardRoom {
    @Override
    public String generateLoot() {
        return "Caves loot x2";
    }

    @Override
    public String generateBoss() {
        return "Chub 300hp";
    }

    @Override
    public String generateSecret(Room neighbour) {
        return "flooded passage after " + neighbour.generateBoss();
    }
}
